### PR TITLE
Planner comparison and update

### PR DESCRIPTION
**Pull request by milroy**
Edit: this PR has grown in scope to enable updates to `planner` and `planner_multi`

To update the Fluxion resource graph, its traverser will need to be reinitialized. The initialization process assembles vectors of resource counts and types via graph traversal to update the `planner_multi` pruning filters. Currently, counts and types are stored as separate vectors in `planner_multi` under the assumption that their order and size don't change. The index of a count must be the same as the index of the corresponding resource type string. This assumption is valid as long as the resource graph is static, but allowing the resource graph to change without updating the planners and their metadata will cause scheduling inconsistency (e.g., incorrect traversal pruning) because of incorrect resource types and counts.

This PR adds comparison operators for fundamental `planner` member data classes to scope the comparison to the correct part of the code. The rescoping avoids relying on awkward class functions for comparison of member data objects and their data.

The PR also restructures `planner_multi` to use `boost::multi_index` container to replace and simplify `planner_multi` member data. The `boost::multi_index` container has the important capability of searching and comparing by multiple indices which can be vector-like (`random_access`), and hash-like (`hashed_unique`). Adding two indexes for lookups enables support for altering resource types, counts, and their orders (indices) in a natural way. Performing these transformations with `vectors` would be higher complexity both in time and lines of code.

Finally, the PR enables the update functionality in the `planner` and `planner_multi` C interfaces and creates unit tests for  updating `planner` and `planner_multi`.

Edit: this PR also re-enables the Valgrind tests and fixes UB in planner unit tests.

### CODE REVIEW

**Review by jameshcorbett — APPROVED** [review]
Were there more changes you wanted to add @milroy ? It looks reasonable to me, although the commits need bodies.

Question though--is the motivation for this mostly to make the code more idiomatic? It doesn't look like it saves much complexity.

**Comment by milroy** [comment]
> Question though--is the motivation for this mostly to make the code more idiomatic? It doesn't look like it saves much complexity.

Yes, being idiomatic and extensible is one motivation for adding the equality operators. The other is that currently the comparisons of `scheduled_point_t` are done via a clunky member function in the `planner` class. Classes like `scheduled_point_t` and `scheduled_point_rb_node_t` should have native comparison operators to avoid scope creep into classes that contain them as member data.

Note that I've significantly added to this PR to add more support for Fluxion resource graph mutability.

**Comment by milroy** [comment]
I re-enabled Valgrind checks and allow `t5000-valgrind.t` to test `planner` and `schema` because the previous `t5000-valgrind.t` didn't detect memory leaks in those Fluxion components. This addresses issue #1136.

File: t/t5000-valgrind.t [hunk 1]
```diff
@@ -48,7 +48,7 @@ test_expect_success \
 	run_timeout 900 \
 	flux start -s ${VALGRIND_NBROKERS} \
 		--killer-timeout=120 \
-		--wrap=libtool,e,${VALGRIND} \
+		--wrap=${VALGRIND} \
```
Comment (jameshcorbett) [inline t/t5000-valgrind.t:51]: Why the change here?

**Review by jameshcorbett — APPROVED** [review]
OK, I _think_ I have a general understanding of what's going on here, and it LGTM. A few comments though.

File: resource/planner/c++/planner.cpp [hunk 2]
```diff
@@ -229,6 +229,26 @@ int planner::restore_track_points ()
...
+int planner::update (uint64_t resource_total)
+{
+    int64_t delta = resource_total - m_total_resources;
+    scheduled_point_t *point = nullptr;
+    if (delta == 0)
+        return 0;
+    m_total_resources = static_cast<int64_t> (resource_total);
+    point = m_sched_point_tree.get_state (m_plan_start);
+    while (point) {
+        // Allow remaining to be negative. This should only happen for
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c++/planner.cpp:241]: If the current value of `point->remaining` is 0 (i.e., the resources are fully allocated/reserved), but `delta` is less than 0 (i.e., a reduction in planner-tracked resources) then the new `point->remaining` value will be negative. I don't foresee problems with this because it should only happen with reservations which will get cleared after a scheduler loop and re-reserved on different resources during the next loop.

If it does cause problems we can just make sure `point->remaining` is strictly nonnegative.

File: resource/planner/c/planner.h [hunk 3]
```diff
@@ -229,6 +229,21 @@ int64_t planner_span_resource_count (planner_t *ctx, int64_t span_id);
...
 
+/*! Update the counts and resource types to support elasticity.
+ *
+ *  \param resource_total
+ *                      64-bit unsigned integer of
+ *                      the total count of available resources
+ *                      of the resource type.
+ *  \param resource_type
+ *                      string containing
+ *                      the planner resource type.
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c/planner.h:240]: Good catch.

File: resource/planner/c++/planner_multi.cpp [hunk 4]
```diff
@@ -86,8 +86,8 @@ planner_multi::planner_multi (const planner_multi &o)
         m_types_totals_planners.push_back ({iter.resource_type,
                                             iter.resource_total, np});
     }
-    m_span_lookup = o.m_span_lookup;
     m_iter = o.m_iter;
+    m_span_lookup = o.m_span_lookup;
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c++/planner_multi.cpp:90]: It's an aesthetic change. I wanted the order of member data copy construction to occur in the same order the member data is defined in the header file.

File: resource/planner/c++/planner_multi.cpp [hunk 5]
```diff
@@ -121,8 +121,8 @@ planner_multi &planner_multi::operator= (const planner_multi &o)
...
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c++/planner_multi.cpp:125]: Same reason as above except in the assignment overload.

File: resource/planner/c++/planner_multi.cpp [hunk 6]
```diff
@@ -184,6 +184,55 @@ planner_multi::~planner_multi ()
...
+    }
+    try {
+        p = new planner_t (base_time, duration,
+                           resource_total,
+                           resource_type);
+    } catch (std::bad_alloc &e) {
+        errno = ENOMEM;
+    }
+    if (p == nullptr)
+        throw std::runtime_error ("ERROR adding planner_multi\n");
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c++/planner_multi.cpp:208]: > Why is there both a catch of std::bad_alloc and a check for p == nullptr? Shouldn't there just be one or the other?

You're right, having both is redundant: https://en.cppreference.com/w/cpp/memory/new/operator_new

> Couldn't maybe both p and type be initialized under the same try {} block?

They could be, but it might be useful information to provide an error message that indicates which generated the `bad_alloc`.

File: resource/planner/test/planner_test01.cpp [hunk 7]
```diff
@@ -639,6 +642,8 @@ static int test_more_add_remove ()
     bo = (bo || span6 == -1);
 
     ok (!bo, "more add-remove-add test works");
+
+    planner_destroy (&ctx);
```
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/test/planner_test01.cpp:646]: That commit is to eliminate a memory leak in the planner test. I could break it into a separate commit.

File: t/t5000-valgrind.t (hunk 1 above)
Comment (milroy) [reply t/t5000-valgrind.t:51]: Since the migration to cmake these arguments break the Valgrind tests. `flux start` exits with an error with those wrap argument.

File: resource/planner/c++/planner_multi.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (milroy) [reply resource/planner/c++/planner_multi.cpp:208]: > You're right, having both is redundant

However, the `planner` copy constructor can throw `std::runtime_error` which will result in `p == nullptr`. In this case I think the code does what we want, but I should change it in the `planner_multi` constructor and for `planner` ctor.

**Comment by milroy** [comment]
Thanks for the feedback @jameshcorbett! I think I addressed your suggestions.

Tagging @trws in case you want to review this, too.

**Comment by zekemorton** [comment]
This looks good to me! Looked through it a couple of times and tested out the branch a bit wit resource query. Nothing jumped out at me as needing to be addressed

**Review by zekemorton — APPROVED** [review]

**Comment by milroy** [comment]
I decided to do a performance test of `hashed_unique` vs `ordered_unique` `boost::multi_index` indexes for creating and querying spans.

Here are the results for `hashed_unique`:
```bash
$ for i in {1..5}; do ./planner_test02 | grep Time; done
Time to add and query 200000 planner_multis: 3.041
Time to add and query 200000 planner_multis: 3.07434
Time to add and query 200000 planner_multis: 3.17595
Time to add and query 200000 planner_multis: 3.06968
Time to add and query 200000 planner_multis: 3.05101
```
in comparison with `ordered_unique`:
```bash
Time to add and query 200000 planner_multis: 3.80358
Time to add and query 200000 planner_multis: 3.82669
Time to add and query 200000 planner_multis: 3.62425
Time to add and query 200000 planner_multis: 3.5208
Time to add and query 200000 planner_multis: 3.15304
```
So I think we should stick with `hashed_unique`.

**Comment by milroy** [comment]
After debugging a segfault in the bookworm clang 15 CI test for quite a long time, I found the culprit. However, I don't understand why it would cause a segfault, or why it would happen only when the unit test is built for clang 15.

This leads to a segfault:

```c++
 bool bo, found = false;
```

While this does not:
```c++
bool bo = false;
bool found = false;
```
Note that the `bool` values aren't related to the segfault, only how `bo` and `found` are declared.
References this commit: 19661cf56359ffdba71abb391409ba6f0d1748a8

@trws do you have any insight here?

**Comment by milroy** [comment]
Deleting a question I asked above because there's a pretty obvious answer that just took a bit of searching.

**Comment by trws** [comment]
Guessing the answer was UB propagation from the undefined value of bo?  I started to type an answer earlier but didn’t get it done before we had to board.  Starting a review.

File: resource/planner/c++/planner.cpp [hunk 8]
```diff
@@ -193,6 +229,26 @@ int planner::restore_track_points ()
     return rc;
 }
 
+int planner::update_total (uint64_t resource_total)
+{
+    int64_t delta = resource_total - m_total_resources;
```
Comment (trws) [inline resource/planner/c++/planner.cpp:234]: Is m_total_resources guaranteed to be smaller than resource_total?

**Review by trws — APPROVED** [review]
Pending the value safety question this looks great.

**Comment by milroy** [comment]
> Guessing the answer was UB propagation from the undefined value of bo?

Yeah, that was it, and I added two commits to fix other occurrences in the unit tests. I thought I was right about the comma-separated syntax for declaration and initialization and went down the wrong investigative hole for a while.

> before we had to board

Thanks for the review and have a good flight!

File: resource/planner/c++/planner.cpp (hunk 8 above)
Comment (milroy) [reply resource/planner/c++/planner.cpp:234]: The case where `resource_total` > `m_total_resources` is when the graph is being grown, and `resource_total` < `m_total_resources` means the graph is being shrunk. There aren't any guarantees on the value of `resource_total` except that it's nonnegative.

@jameshcorbett asked a form of this question, too, and my reasoning has been that negative values for `point->remaining` (e.g., when `(point->remaining + delta) < 0`) shouldn't cause problems because they should only happen for reservations (which will be cleared at the next schedule loop). I was thinking that adding a check for negative remaining wasn't worth the performance penalty.

That said, I don't expect `update_total` to be called often, and adding a guard check may prevent complicated problems with removing spans if that process isn't carefully performed upstream (i.e., in `planner_c_interface`).

**Comment by milroy** [comment]
Thanks for the reviews all! Setting MWP.
